freqtrade/freqai/RL/Base5ActionRLEnv.py

```diff
@@ -0,0 +1,201 @@
+                trade_type = "short"
+                self._last_trade_tick = self._current_tick
+            elif action == Actions.Long_exit.value:
+                self._update_total_profit()
+                self._position = Positions.Neutral
+                trade_type = "neutral"
+                self._last_trade_tick = None
+            elif action == Actions.Short_exit.value:
+                self._update_total_profit()
+                self._position = Positions.Neutral
+                trade_type = "neutral"
+                self._last_trade_tick = None
+            else:
+                print("case not defined")
+
+            if trade_type is not None:
+                self.trade_history.append(
+                    {'price': self.current_price(), 'index': self._current_tick,
+                     'type': trade_type})
+
+        if (self._total_profit < self.max_drawdown or
+                self._total_unrealized_profit < self.max_drawdown):
+            self._done = True
+
+        self._position_history.append(self._position)
+
+        info = dict(
+            tick=self._current_tick,
+            total_reward=self.total_reward,
+            total_profit=self._total_profit,
+            position=self._position.value
+        )
+
+        observation = self._get_observation()
+
+        self._update_history(info)
+
+        return observation, step_reward, self._done, info
+
+    def _get_observation(self):
+        features_window = self.signal_features[(
+            self._current_tick - self.window_size):self._current_tick]
+        features_and_state = DataFrame(np.zeros((len(features_window), 3)),
+                                       columns=['current_profit_pct', 'position', 'trade_duration'],
+                                       index=features_window.index)
+
+        features_and_state['current_profit_pct'] = self.get_unrealized_profit()
+        features_and_state['position'] = self._position.value
+        features_and_state['trade_duration'] = self.get_trade_duration()
+        features_and_state = pd.concat([features_window, features_and_state], axis=1)
+        return features_and_state
+
+    def get_trade_duration(self):
+        if self._last_trade_tick is None:
+            return 0
+        else:
+            return self._current_tick - self._last_trade_tick
+
+    def is_tradesignal(self, action: int):
+        # trade signal
+        """
+        Determine if the signal is a trade signal
+        e.g.: agent wants a Actions.Long_exit while it is in a Positions.short
+        """
+        return not ((action == Actions.Neutral.value and self._position == Positions.Neutral) or
+                    (action == Actions.Neutral.value and self._position == Positions.Short) or
+                    (action == Actions.Neutral.value and self._position == Positions.Long) or
+                    (action == Actions.Short_enter.value and self._position == Positions.Short) or
```

Comment:
It can be only valid, if the user wants use DCA, it is also a possibility, but yes you are right, the DCA has not set into example strategy or the environment as a possibility.

Reply:
be careful with "advanced" functionality (it's also wrong btw) in this environment.
it's current implementation is ONLY respecting buy/sell signals (no ROI, no trailing stop, no advanced functionality like callbacks).
now i'm not sure all of the above would make sense from a performance perspective (it can be quite expensive to have to call a callback for every candle).

a regular additional enter/exit signal would never cause a DCA action - that'll come from the callback - which could have completely different conditions.

Reply:
Btw, this function goes to "False"  if it get Short_enter in a Short Position, because it is  `return not` ! So it is totally right and following the example basic strategy

Reply:
Yeah I don't see the issue here, it is saying "if its in a short and it also has a short_enter, then return a False for `is_tradesignal`. i.e., the function says "that is not a valid trade signal.